### PR TITLE
dialects: (builtin) Fix printing and parsing of `UnitAttr`

### PR DESCRIPTION
**Pull request by superlopuh**
It's actually `unit` not the empty string.

### CODE REVIEW

File: xdsl/parser/attribute_parser.py [hunk 1]
```diff
@@ -1250,6 +1250,7 @@ def _parse_optional_integer_or_float_type(self) -> Attribute | None:
         Parse as integer or float type, if present.
           integer-or-float-type ::= index-type | integer-type | float-type
           index-type            ::= `index`
+          unit-attribute        ::= `unit`
```
Comment (superlopuh) [inline xdsl/parser/attribute_parser.py:1253]: I first tried to add this as a normal attribute not type, but then the test failed, saying "type expected". I have a supsicion it's actually a type in MLIR also, but my quick search for the definition didn't yield much so I thought I'd open a PR anyway.

Comment (math-fehr) [reply xdsl/parser/attribute_parser.py:1253]: It is not a type, it's an attribute though: https://mlir.llvm.org/docs/Dialects/Builtin/#unitattr

Comment (superlopuh) [reply xdsl/parser/attribute_parser.py:1253]: Should I change the test instead? When I put the code for parsing into `parse_attribute` it fails, as the block argument expects a type.

File: xdsl/parser/attribute_parser.py [hunk 2]
```diff
@@ -1262,6 +1263,11 @@ def _parse_optional_integer_or_float_type(self) -> Attribute | None:
             self._consume_token()
             return IndexType()
 
+        # Unit type
+        if name == "unit":
+            self._consume_token()
+            return UnitAttr()
+
```
Comment (math-fehr) [inline xdsl/parser/attribute_parser.py:1270]: This is not the right place to add this though, this is neither an integer nor attribute type?

**Review by math-fehr — COMMENTED** [review]
Thanks for looking at this!
`unit` is an attribute, not a type in MLIR, and we should stay the same.
So here I would not add it in the `parse_optional_integer_or_float_type`

File: xdsl/parser/attribute_parser.py (hunk 1 above)
Comment (math-fehr) [reply xdsl/parser/attribute_parser.py]: Which test? And yes, we should change it I think

File: tests/filecheck/parser-printer/builtin_attrs.mlir [hunk 3]
```diff
@@ -8,6 +8,9 @@
 
   // CHECK: (index)
 
+  "func.func"() ({}) {function_type = () -> (), sym_name = "unit_attr_func", unitarray = [unit]} : () -> ()
```
Comment (AntonLydike) [inline tests/filecheck/parser-printer/builtin_attrs.mlir:11]: Can I suggest using the `test` dialect here? I don't think we want to depend on `func` here...

File: xdsl/parser/attribute_parser.py [hunk 4]
```diff
@@ -651,6 +652,22 @@ def _parse_strided_layout_attr(self, name: Span) -> Attribute:
...
         return StridedLayoutAttr(strides, None if offset == "?" else offset)
 
+    def parse_optional_unit_attr(self) -> Attribute | None:
+        """
+        Parse a value of `unit` type.
+        unit-attribute ::= `unit`
+        """
+        if self._current_token.kind != Token.Kind.BARE_IDENT:
+            return None
+        name = self._current_token.span.text
```
Comment (AntonLydike) [inline xdsl/parser/attribute_parser.py:662]: supernit: Not sure this extra variable assignment is needed here?

**Review by AntonLydike — APPROVED** [review]
LGTM, just some small misc. comments

**Review by georgebisbas — APPROVED** [review]

File: tests/filecheck/parser-printer/builtin_attrs.mlir (hunk 3 above)
Comment (superlopuh) [reply tests/filecheck/parser-printer/builtin_attrs.mlir:11]: I agree in general, although here it seemed like a good fit for consistency, happy to change it if you or anyone else insist

File: xdsl/parser/attribute_parser.py (hunk 4 above)
Comment (superlopuh) [reply xdsl/parser/attribute_parser.py:662]: definitely not, although it doesn't really hurt either

File: xdsl/parser/attribute_parser.py [hunk 5]
```diff
@@ -651,6 +652,22 @@ def _parse_strided_layout_attr(self, name: Span) -> Attribute:
...
+
+        # Unit type
```
Comment (PapyChacal) [inline xdsl/parser/attribute_parser.py:664]:
```suggestion
        # Unit attribute
```

**Review by PapyChacal — APPROVED** [review]
Seeing the low footprint, I'm guessing this preserves the behaviour of unit attribute being printed as "just a name" in attribute dictionaries?

I had no clue they were handled otherwise in other places 😃

**Review by math-fehr — APPROVED** [review]
Looks good to me now (I would just fix the comment that Emilien pointed)

**Pull request merged**